text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+};
+```
+
+## Default fields values are [`const` context]s
+
+[`const` context]: https://github.com/rust-lang-nursery/reference/blob/66ef5396eccca909536b91cad853f727789c8ebe/src/const_eval.md#const-context
+
+As you saw in (7), `Vec::new()`, a function call, was used.
+However, this assumes that `Vec::new` is a *`const fn`*. That is, when you
+provide a default value `field: Type = value`, the given `value` must be a
+*constant expression* such that it is valid in a [`const` context].
+Therefore, you cannot write something like (10):
+
+```rust
+fn launch_missilies() -> Result<(), LaunchFailure> {
+    authenticate()?;
+    begin_launch_sequence()?;
+    ignite()?;
+    Ok(())
+}
+
+struct BadFoo {
+    bad_field: u8 = {
+        launch_missilies().unwrap();
+        42
+    },
+}
+```
+
+Since launching missiles interacts with the real world and has *side-effects*
+in it, it is not possible to do that in a `const` context since it may violate
+deterministic compilation.
+
+## Privacy interactions
+
+In examples thus far, initialization literal expressions `Foo { bar, .. }` have
+been used in the same module as where the constructed type was defined in.
+
+Let's now consider a scenario where this is not the case (11):
+
+```rust
+pub mod foo {
+    pub struct Alpha {
+        beta: u8 = 42,
+        gamma: bool = true,
+    }
+}
+
+mod bar {
+    fn baz() {
+        let x = Alpha { .. };
+    }
+}
+```
+
+Despite `foo::bar` being in a different module than `foo::Alpha` and despite
+`beta` and `gamma` being private to `foo::bar`, a Rust compiler will accept
+the above snippet. It is legal because when `Alpha { .. }` expands to
+`Alpha { beta: 42, gamma: true }`, the fields `beta` and `gamma` are considered
+in the context of `foo::Alpha`'s *definition site* rather than `bar::baz`'s
+definition site.
+
+## `#[non_exhaustive]` interactions
+
+[RFC 2008]: https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#structs-1
+
+[RFC 2008] introduced the attribute `#[non_exhaustive]` that can be placed
+on `struct`, `enum`, and `enum` variants. The RFC notes that upon defining
+a `struct` in *crate A* such as (12):
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub width: u16,
+    pub height: u16,
+}
+```
+
+it is **_not_** possible to initialize a `Config` in a different *crate B* (13):
+
+```rust
+let config = Config { width: 640, height: 480 };
+```
+
+This is forbidden when `#[non_exhaustive]` is attached because the purpose of
+the attribute is to permit adding fields to `Config` without causing a
+breaking change. However, the RFC goes on to note that you can pattern match
+if you allow for the possibility of having fields be ignored with `..` (14):
+
+```rust
+let Config { width, height, .. } = config;
+```
+
+Because this RFC gives a way to have default field values, you can now simply
+invert the pattern expression and initialize a `Config` like so (15):
+
+```rust
+let config = Config { width, height, .. };
+```
```

Comment:
Just to see if I get you right, you saying that this RFC should *disallow* `#[non_exhaustive]` *and* default fields on the same struct, and leave any interaction as an open question? I'm down with that.

Reply:
Honestly, I think it would be useful to preserve the property that `Struct { field: value, .. }` as an expression (the proposed syntax) is equivalent to `Struct { field: value, ..Default::default() }`, and as such, these examples would only work if these structs derived or manually implemented `Default`.

That should cover the API concerns, although it would make `Default` become a land item, which I am personally fine with but I am not everyone, so, that would be a downside to this approach.

If it doesn't interact with `Default` at all, I agree that it shouldn't allow this, since it does break APIs.

Reply:
I think requiring `Default` is unnecessarily limiting, since it would prevent using the nice new syntax with structs where some fields have defaults and others intentionally do not, e.g. if `syn::Index` (a tuple struct field name) used the new syntax it could be:
```rust
pub struct Index {
    pub index: u32,
    pub span: Span = Span::call_site(),
}
```

Reply:
Rereading this I believe that the part of the feature that has a bad interaction is allowing the construction of types that have private default field values. Disallowing that in general, or when the type is *also* annotated with `non_exhaustive` would eliminate the issue for every cross-crate case, right? The only potential issue would be intra-crate, and that's already out of scope of `non_exhaustive`, IIUC.

Reply:
On reading the RFC, I was honestly a little confused on why interaction with non_exhaustive would be a problem.

My understanding is that non_exhaustive means you can't construct the struct outside of the crate. The ability to have defaults shouldn't have any impact on that. And prohibiting use of defaults on a non_exhaustive struct seems unnecessarily restrictive, since that could still be useful within the crate.

Now, there may be value in a non_exhaustive_but_future_fields_have_defaults functionality, but I think that should be a separate attribute, or add an argument to the existing non_exhaustive attribute, not usurp and change the meaning of the current non_exhaustive attribute.